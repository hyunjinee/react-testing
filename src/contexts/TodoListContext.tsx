import { createContext, useEffect, useState } from 'react';

interface Context {
  readonly todoList: string[];
  readonly addTodo: (todo: string) => void;
  readonly deleteTodo: (index: number) => void;
}

export const TodoListContext = createContext<Context>({
  todoList: [],
  addTodo: () => {},
  deleteTodo: () => {},
});

interface Props {
  children: JSX.Element | JSX.Element[];
}

export const TodoListProvider: React.FC<Props> = ({ children }) => {
  const [todoList, setTodoList] = useState<string[]>([]);

  useEffect(() => {
    const list = localStorage.getItem('todoList');
    if (list) {
      setTodoList(JSON.parse(list));
    }
  }, []);

  const addTodo = (todo: string) => {
    if (todo) {
      const newList = [...todoList, todo];
      localStorage.setItem('todoList', JSON.stringify(newList));
      setTodoList(newList);
    }
  };

  const deleteTodo = (index: number) => {
    let list = [...todoList];
    list.splice(index, 1);
    localStorage.setItem('todoList', JSON.stringify(list));
    setTodoList(list);
  };

  return (
    <TodoListContext.Provider
      value={{
        todoList,
        addTodo,
        deleteTodo,
      }}
    >
      {children}
    </TodoListContext.Provider>
  );
};
