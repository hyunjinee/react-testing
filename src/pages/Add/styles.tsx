import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  background-color: #ffffff;

  padding: 20px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
  border-radius: 8px;
  position: relative;
  align-items: center;
`;
