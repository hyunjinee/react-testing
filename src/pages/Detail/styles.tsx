import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  background-color: 'fff';
  padding: 20px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
  border-radius: 8px;
  align-items: center;
  flex-direction: column;
`;

export const Todo = styled.div`
  min-width: 350px;
  height: 350px;
  overflow-y: auto;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
  padding: 10px;
`;
