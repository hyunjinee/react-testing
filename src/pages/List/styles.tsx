import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  background-color: #ffffff;

  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
  position: relative;
  padding: 20px;
  align-items: center;
`;

export const AddButton = styled(Link)`
  font-size: 20px;
  color: #ffffff;
  background-color: #304ffe;
  width: 60px;
  height: 60px;
  border-radius: 50%;
  box-shadow: 5px 5px 10px rgba(0, 0, 0, 0.2);
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  position: absolute;
  bottom: -30px;
  text-decoration: none;

  &:hover {
    background-color: #1e40ff;
  }
  &:active {
    box-shadow: inset 5px 5px 10px rgba(0, 0, 0, 0.2);
  }
`;
