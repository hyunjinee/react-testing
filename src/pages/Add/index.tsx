import InputContainer from 'components/InputContainer';
import { useNavigate } from 'react-router-dom';
import { Container } from './styles';

const Add: React.FC = () => {
  const navigate = useNavigate();

  return (
    <Container>
      <InputContainer onAdd={() => navigate('/', { replace: true })} />
    </Container>
  );
};

export default Add;
