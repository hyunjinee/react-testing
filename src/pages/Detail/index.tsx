import { useContext } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { Container, Todo } from './styles';
import { TodoListContext } from 'contexts/TodoListContext';
import Button from 'components/common/Button';

const Detail: React.FC = () => {
  const { id } = useParams<{ id: string }>();
  const { todoList, deleteTodo } = useContext(TodoListContext);
  const navigate = useNavigate();

  const todo = todoList[parseInt(id as string)];

  return (
    <Container>
      <Todo>{todo}</Todo>
      <Button
        label="삭제"
        backgroundColor="#ff1744"
        hoverColor="#f01440"
        onClick={() => {
          deleteTodo(parseInt(id as string));
          navigate(-1);
        }}
      />
    </Container>
  );
};

export default Detail;
