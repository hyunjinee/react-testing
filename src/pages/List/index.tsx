import TodoList from 'components/TodoList';
import { AddButton, Container } from './styles';

const List: React.FC = () => {
  return (
    <Container>
      <TodoList />
      <AddButton to="/add">+</AddButton>
    </Container>
  );
};

export default List;
