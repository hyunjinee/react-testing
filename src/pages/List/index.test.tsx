import { MemoryRouter, Route, Routes, useLocation } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { TodoListProvider } from 'contexts/TodoListContext';
import Detail from 'pages/Detail';
import List from '.';
import Add from 'pages/Add';

describe('<List/>', () => {
  it('renders component correctly', () => {
    localStorage.setItem('todoList', '["할 일 1", "할 일 2"]');

    const { container } = render(
      <TodoListProvider>
        <MemoryRouter>
          <Routes>
            <Route path="/" element={<List />} />
            <Route path="/detail/:id" element={<Detail />} />
          </Routes>
        </MemoryRouter>
      </TodoListProvider>,
    );

    const todoItem = screen.getByText('할 일 1');
    expect(todoItem).toBeInTheDocument();
    expect(todoItem.getAttribute('href')).toBe('/detail/0');

    const todoItem2 = screen.getByText('할 일 2');
    expect(todoItem2).toBeInTheDocument();
    expect(todoItem2.getAttribute('href')).toBe('/detail/1');

    expect(screen.getAllByText('삭제')).toHaveLength(2);

    const addButton = screen.getByText('+');
    expect(addButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('deletes todo item', () => {
    localStorage.setItem('todoList', '["할 일 1", "할 일 2"]');

    render(
      <TodoListProvider>
        <MemoryRouter>
          <Routes>
            <Route path="/" element={<List />} />
            <Route path="/detail/:id" element={<Detail />} />
          </Routes>
        </MemoryRouter>
      </TodoListProvider>,
    );

    const todoItem = screen.getByText('할 일 1');
    expect(todoItem).toBeInTheDocument();
    // eslint-disable-next-line testing-library/no-node-access
    fireEvent.click(todoItem.nextElementSibling as HTMLElement);
    expect(todoItem).not.toBeInTheDocument();
    expect(
      JSON.parse(localStorage.getItem('todoList') as string),
    ).toStrictEqual(['할 일 2']);
  });

  it('moves to Detail page', () => {
    const TestComponent = (): JSX.Element => {
      const { pathname } = useLocation();
      return <div>{pathname}</div>;
    };

    localStorage.setItem('todoList', '["할 일 1", "할 일 2"]');

    render(
      <TodoListProvider>
        <MemoryRouter>
          <TestComponent />
          <Routes>
            <Route path="/" element={<List />} />
            <Route path="/detail/:id" element={<Detail />} />
          </Routes>
        </MemoryRouter>
      </TodoListProvider>,
    );

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();

    const todoItem = screen.getByText('할 일 1');
    expect(todoItem.getAttribute('href')).toBe('/detail/0');
    fireEvent.click(todoItem);

    expect(url.textContent).toBe('/detail/0');
  });

  it('moves to add page', () => {
    const TestComponent = (): JSX.Element => {
      const { pathname } = useLocation();
      return <div>{pathname}</div>;
    };

    render(
      <MemoryRouter>
        <TestComponent />
        <Routes>
          <Route path="/" element={<List />} />
          <Route path="/add" element={<Add />} />
        </Routes>
      </MemoryRouter>,
    );

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();

    const addButton = screen.getByText('+');
    fireEvent.click(addButton);

    expect(url.textContent).toBe('/add');
  });
});
