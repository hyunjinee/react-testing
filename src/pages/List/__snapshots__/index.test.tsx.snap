// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List/> renders component correctly 1`] = `
<div>
  <div
    class="sc-iBkjds KavGT"
  >
    <div
      class="sc-gKXOVf eJVsLV"
      data-testid="todoList"
    >
      <div
        class="sc-eCYdqJ fHVchU"
      >
        <a
          class="sc-jSMfEi fzvlKC"
          href="/detail/0"
        >
          할 일 1
        </a>
        <div
          class="sc-dkzDqf dgyiws"
        >
          <div
            class="sc-hKMtZM dTJMco"
          >
            삭제
          </div>
        </div>
      </div>
      <div
        class="sc-eCYdqJ fHVchU"
      >
        <a
          class="sc-jSMfEi fzvlKC"
          href="/detail/1"
        >
          할 일 2
        </a>
        <div
          class="sc-dkzDqf dgyiws"
        >
          <div
            class="sc-hKMtZM dTJMco"
          >
            삭제
          </div>
        </div>
      </div>
    </div>
    <a
      class="sc-ftvSup kUpiXW"
      href="/add"
    >
      +
    </a>
  </div>
</div>
`;
